feat(chart): add support for  custom TLS secret for api.ingress, fallback to self-signed

abelhoula opened the pull request:
## Description

Add per-resource helpers in `charts/kargo/templates/_tls.tpl` to resolve TLS Secret names
with a single, explicit rule:

  If `selfSignedCert=false` AND `secretName` is non-empty -> use `secretName`
  Else -> use the component’s default self-signed Secret name

Applied uniformly to all six TLS-capable locations:

1) API server TLS                          -> kargo-api-cert
2) API Ingress TLS                         -> kargo-api-ingress-cert
3) Dex server TLS                          -> kargo-dex-server-cert
4) External Webhooks Server TLS            -> kargo-external-webhook-server-cert
5) External Webhooks Server Ingress TLS    -> kargo-external-webhook-server-ingress-cert
6) Built-in Webhooks Server TLS            -> kargo-webhooks-server-cert

Also:
- Clarify inline `values.yaml` docs to state that `secretName` is only honored
  when `selfSignedCert=false`; otherwise it’s ignored.
- Keep existing defaults and behavior when `secretName` is not provided or when
  `selfSignedCert=true`.

## Motivation:
Some environments cannot control Secret naming conventions and need to supply
an existing Secret whose name does not match the chart defaults. This change
enables that, without breaking existing flows.

## Backwards compatibility:
- No change for existing users that rely on self-signed certs (default).
- Users providing BYO secrets can now set `selfSignedCert=false` + `secretName`
  to have the chart pick up their Secret.
- No fails/fatal validations added; behavior is purely precedence.

## Testing:
- Verified with a minimal overrides file exercising all six code paths; each
  rendered the configured `secretName` when `selfSignedCert=false`.

abelhoula commented:
Many platforms pre-provision TLS secrets with standardized names (e.g., from external PKI/ACME controllers). Allowing `secretName `lets us point to those directly without renaming or duplicating secrets.

krancour commented:
I think the case wherein one has no control over the naming convention of the Secrets containing one's certs, outlying though I think it may be, seems like rational impetus for this.

So I'm on board with the idea, but the following things need to be addressed for this PR to be merged:

1. I think there is a total of six places where you can either use a self-signed cert or bring your own. For consistency's sake, this change should be applied consistently in all six places instead of just the one place where you care about it.

2. I think some more discussion is in order over how exactly this is to work and needs to be documented more thoroughly. The following bit of inline doc in `values.yaml` leaves me with questions that I need to go digging in the templates to answer:

    > Name of the existing TLS secret to use for the API server's Ingress.

    It's not clear from that what relationship, if any, it has to `selfSignedCert`. Is it applicable only when `selfSignedCert` is `false`? This does not appear to be the case, but I think it ought to be. What happens if `selfSignedCert` is true and value is specified for `secretName` regardless? The current implementation doesn't appear to deal with that scenario at all.

Bottom line is that I am perfectly amenable to this change, but need it to be more thorough.

abelhoula commented:
@krancour Thanks for the thoughtful feedback! I’ve addressed everything you called out—details are in the updated PR description, but the short version:

Applied consistently in all six places
Introduced per-resource helpers in templates/_tls.tpl and wired them everywhere TLS can be self-signed or BYO:

1. API server TLS → kargo.api.serverTLSSecretName
2. API Ingress TLS → kargo.api.ingressTLSSecretName
3. Dex server TLS → kargo.dex.serverTLSSecretName
4. External Webhooks Server TLS → kargo.externalWebhooks.serverTLSSecretName
5. External Webhooks Server Ingress TLS → kargo.externalWebhooks.ingressTLSSecretName
6. Built-in Webhooks Server TLS → kargo.webhooksServer.tlsSecretName

krancour commented on charts/kargo/values.yaml:
```diff
@@ -301,7 +302,8 @@ api:
 
       ## @param api.oidc.dex.resources Resources limits and requests for the Dex server containers.
-      resources: {}
+      resources:
+        {}
```
```suggestion
      resources: {}
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -382,7 +387,8 @@ api:
 
   ## @param api.resources Resources limits and requests for the api containers.
-  resources: {}
+  resources:
+    {}
```
```suggestion
  resources: {}
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -414,14 +420,18 @@ api:
     enabled: false
     ## @param api.ingress.annotations Annotations specified by your ingress controller to customize the behavior of the Ingress resource.
-    annotations: {}
+    annotations:
+      {}
```
```suggestion
    annotations: {}
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -563,7 +576,8 @@ controller:
 
   ## @param controller.resources Resources limits and requests for the controller containers.
-  resources: {}
+  resources:
+    {}
```
```suggestion
  resources: {}
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -813,7 +836,8 @@ managementController:
 
   ## @param managementController.resources Resources limits and requests for the management controller containers.
-  resources: {}
+  resources:
+    {}
```
```suggestion
  resources: {}
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -414,6 +415,9 @@ api:
     enabled: true
     ## @param api.tls.selfSignedCert Whether to generate a self-signed certificate for use by the API server. If `true`, `cert-manager` CRDs **must** be present in the cluster. Kargo will create and use its own namespaced issuer. If `false`, a cert secret named `kargo-api-cert` **must** be provided in the same namespace as Kargo.
     selfSignedCert: true
+    ## @param api.tls.secretName Name of the existing TLS secret to use for the API server's.
+    ## Required when `api.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```
```suggestion
    ## @param api.tls.secretName Name of the existing TLS secret to use for the API server's. Required when `api.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -728,6 +735,9 @@ externalWebhooksServer:
     enabled: true
     ## @param externalWebhooksServer.tls.selfSignedCert Whether to generate a self-signed certificate for use by the external webhooks server. If `true`, `cert-manager` CRDs **must** be present in the cluster. Kargo will create and use its own namespaced issuer. If `false`, a cert secret named `kargo-external-webhook-server-cert` **must** be provided in the same namespace as Kargo.
     selfSignedCert: true
+    ## @param externalWebhooksServer.tls.secretName Name of the existing TLS secret for the external webhooks servers.
+    ## Required when `externalWebhooksServer.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```
```suggestion
    ## @param externalWebhooksServer.tls.secretName Name of the existing TLS secret for the external webhooks servers. Required when `externalWebhooksServer.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```

krancour commented on charts/kargo/values.yaml:
```diff
@@ -744,6 +754,9 @@ externalWebhooksServer:
       enabled: true
       ## @param externalWebhooksServer.ingress.tls.selfSignedCert Whether to generate a self-signed certificate for use with the external webhook server's Ingress resource. If `true`, `cert-manager` CRDs **must** be present in the cluster. Kargo will create and use its own namespaced issuer. If `false`, a cert secret named `kargo-external-webhook-server-ingress-cert` **must** be provided in the s ...
       selfSignedCert: true
+      ## @param externalWebhooksServer.ingress.tls.secretName Name of the existing TLS secret for the external webhooks servers Ingress resource.
+      ## Required when `externalWebhooksServer.ingress.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```
```suggestion
      ## @param externalWebhooksServer.ingress.tls.secretName Name of the existing TLS secret for the external webhooks servers Ingress resource. Required when `externalWebhooksServer.ingress.tls.selfSignedCert` is `false`. Ignored if `selfSignedCert` is `true`.
```

krancour commented on charts/kargo/templates/api/ingress.yaml:
```diff
@@ -47,8 +47,8 @@ spec:
-  - hosts:
-    - {{ quote .Values.api.host }}
-    secretName: kargo-api-ingress-cert
+    - hosts:
+        - {{ quote .Values.api.host }}
+      secretName: {{ include "kargo.api.ingressTLSSecretName" . | quote }}
```
```suggestion
  - hosts:
    - {{ quote .Values.api.host }}
    secretName: {{ include "kargo.api.ingressTLSSecretName" . | quote }}
```

krancour commented:
The helpers seem like an unnecessary complication. It remain non-breaking, but would also be much more straightforward if `secretName` just had a default value equal to the current hard-coded name. When certs are self-signed, that will be the name of the Secret generated and when they're not, that will be the name of the Secret you must provide.

I've refactored in this way and thoroughly tested it.
